restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java

```diff
@@ -397,6 +404,110 @@ private Object convertToBackendBooleanValue(boolean value, boolean numericBoolea
     return ImmutablePair.of(bindVariableList, firstLevelKeys);
   }
 
+  private Optional<String> convertToJsonPtr(Optional<String> path) {
+    return path.map(
+        p -> "/" + p.replaceAll(PERIOD_PATTERN.pattern(), "/").replaceAll("\\[(\\d+)\\]", "$1"));
+  }
+
+  private DocumentDB maybeCreateTableAndIndexes(
+      Db dbFactory,
+      DocumentDB db,
+      String keyspace,
+      String collection,
+      Map<String, String> headers,
+      String authToken)
+      throws UnauthorizedException {
+    boolean created = db.maybeCreateTable(keyspace, collection);
+    // After creating the table, it can take up to 2 seconds for permissions cache to be updated,
+    // but we can force the permissions refetch by logging in again.
+    if (created) {
+      db = dbFactory.getDocDataStoreForToken(authToken, headers);
+      db.maybeCreateTableIndexes(keyspace, collection);
+    }
+    return db;
+  }
+
+  public List<String> writeManyDocs(
+      String authToken,
+      String keyspace,
+      String collection,
+      InputStream payload,
+      Optional<String> idPath,
+      Db dbFactory,
+      ExecutionContext context,
+      Map<String, String> headers)
+      throws IOException, UnauthorizedException {
+
+    DocumentDB db = dbFactory.getDocDataStoreForToken(authToken, headers);
+    JsonSurfer surfer = JsonSurferGson.INSTANCE;
+
+    db = maybeCreateTableAndIndexes(dbFactory, db, keyspace, collection, headers, authToken);
+    List<String> idsWritten = new ArrayList<>();
+    try (JsonParser jsonParser = mapper.getFactory().createParser(payload)) {
+      Optional<String> docsPath = convertToJsonPtr(idPath);
+
+      Map<String, String> docs = new LinkedHashMap<>();
+      if (jsonParser.nextToken() != JsonToken.START_ARRAY) {
+        throw new IllegalArgumentException("Payload must be an array.");
+      }
+
+      while (jsonParser.nextToken() != JsonToken.END_ARRAY) {
+        JsonNode json = mapper.readTree(jsonParser);
+        String docId;
+        if (idPath.isPresent()) {
```

Comment:
shouldn't this be `docsPath.isPresent()`?

Reply:
technically it is logically equivalent, but yea that would make way more sense to read :) I'll change

Reply:
Wondering if we could hit a `JsonToken.END_ARRAY` inside a document itself and loosing track witch the streaming parser here?

```json
[
 { "id":1, "sample":[1,2,3] },
 { "id":2, "sample":[4,5,6] }
]
```

Reply:
I've tested this out, and the parser actually works properly! This is because mapper.readTree(...) on the parser reads the next full object.

Reply:
See https://github.com/stargate/stargate/pull/1043/files#diff-585290b1d8a4cd9f6b594dd2e6a35102c4ab7c86988b9df67452177d38c73868R1, which is used in a few tests